worker/src/main.rs

```diff
@@ -524,44 +524,20 @@ pub fn produce_blocks<E: EnclaveBase + SideChain>(
 	debug!("Getting current head");
 	let curr_head: SignedBlock = api.last_finalized_block().unwrap().unwrap();
 
-	let mut blocks_to_sync = Vec::<SignedBlock>::new();
-
-	// add blocks to sync if not already up to date
-	if curr_head.block.header.hash() != last_synced_head.hash() {
-		blocks_to_sync.push(curr_head.clone());
-
-		// Todo: Check, is this dangerous such that it could be an eternal or too big loop?
-		let mut head = curr_head.clone();
-		let no_blocks_to_sync = head.block.header.number - last_synced_head.number;
-		if no_blocks_to_sync > 1 {
-			println!("Chain Relay is synced until block: {:?}", last_synced_head.number);
-			println!("Last finalized block number: {:?}\n", head.block.header.number);
-		}
-		while head.block.header.parent_hash != last_synced_head.hash() {
-			debug!("Getting head of hash: {:?}", head.block.header.parent_hash);
-			head = api.signed_block(Some(head.block.header.parent_hash)).unwrap().unwrap();
-			blocks_to_sync.push(head.clone());
-
-			if head.block.header.number % BLOCK_SYNC_BATCH_SIZE == 0 {
-				println!(
-					"Remaining blocks to fetch until last synced header: {:?}",
-					head.block.header.number - last_synced_head.number
-				)
-			}
-		}
-		blocks_to_sync.reverse();
-	}
+	let blocks_to_sync = get_blocks_to_sync(api, &last_synced_head, &curr_head);
 
 	let tee_accountid = enclave_account(enclave_api);
 
 	// only feed BLOCK_SYNC_BATCH_SIZE blocks at a time into the enclave to save enclave state regularly
-	let mut i = if curr_head.block.header.hash() == last_synced_head.hash() {
+	let mut current_block_number = if curr_head.block.header.hash() == last_synced_head.hash() {
 		curr_head.block.header.number as usize
 	} else {
-		blocks_to_sync[0].block.header.number as usize
+		blocks_to_sync.first().map(|h| h.block.header.number as usize).unwrap_or(0)
```

Comment:
potential panic if we just 'blindly' access element `[0]` - replaced by `first()` and subsequent handling of the `None` case.

Reply:
Good job detecting that. (The above logic probably made sure that this is never empty when we reach here, but it is still much better to prevent that panic here at all costs.)